File: scripts/ci/check-disk-headroom.mjs
#!/usr/bin/env node

import { statfsSync } from 'node:fs'
import process from 'node:process'
import { pathToFileURL } from 'node:url'

const GIB = 1024n ** 3n
const DEFAULT_MINIMUM_GIB = 10

export function evaluateHeadroom(freeBytes, minimumBytes) {
  if (freeBytes >= minimumBytes) return null
  const freeGiB = Number(freeBytes) / Number(GIB)
  const minimumGiB = Number(minimumBytes) / Number(GIB)
  return (
    `capacity preflight failed before ENOSPC: ${freeGiB.toFixed(2)} GiB available; ` +
    `${minimumGiB.toFixed(2)} GiB required. Remove dependency trees only from clean ` +
    'completed worktrees, prune completed worktrees after preserving their refs, or ' +
    'clear rebuildable package caches. Never clean active workspaces or deployment state.'
  )
}

function parseArgs(argv) {
  let minimumGiB = DEFAULT_MINIMUM_GIB
  let path = process.cwd()
  for (let index = 0; index < argv.length; index += 1) {
    if (argv[index] === '--minimum-gib') {
      minimumGiB = Number(argv[++index])
    } else if (argv[index] === '--path') {
      path = argv[++index]
    } else {
      throw new Error(`unknown argument: ${argv[index]}`)
    }
  }
  if (!Number.isFinite(minimumGiB) || minimumGiB <= 0) {
    throw new Error('--minimum-gib must be greater than zero')
  }
  return { minimumGiB, path }
}

export function main(argv = process.argv.slice(2)) {
  const { minimumGiB, path } = parseArgs(argv)
  const stats = statfsSync(path, { bigint: true })
  const freeBytes = stats.bavail * stats.bsize
  const minimumBytes = BigInt(Math.ceil(minimumGiB * 1024 ** 3))
  const error = evaluateHeadroom(freeBytes, minimumBytes)
  if (error) {
    console.error(`error: ${error}`)
    return 1
  }
  console.log(
    `capacity preflight passed: ${(Number(freeBytes) / Number(GIB)).toFixed(2)} GiB available; ` +
      `${minimumGiB.toFixed(2)} GiB required`,
  )
  return 0
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  try {
    process.exitCode = main()
  } catch (error) {
    console.error(`error: ${error.message}`)
    process.exitCode = 2
  }
}

File: scripts/ci/check_disk_headroom.py
#!/usr/bin/env python3
"""Fail local integration work before low disk headroom becomes ENOSPC."""

from __future__ import annotations

import argparse
import math
import shutil
from pathlib import Path

GIB = 1024**3
DEFAULT_MINIMUM_GIB = 10.0


def evaluate_headroom(free_bytes: int, minimum_bytes: int) -> str | None:
    """Return an actionable error when free space is below the floor."""
    if free_bytes >= minimum_bytes:
        return None
    free_gib = free_bytes / GIB
    minimum_gib = minimum_bytes / GIB
    return (
        "capacity preflight failed before ENOSPC: "
        f"{free_gib:.2f} GiB available; {minimum_gib:.2f} GiB required. "
        "Remove dependency trees only from clean completed worktrees, prune "
        "completed worktrees after preserving their refs, or clear rebuildable "
        "package caches. Never clean active workspaces or deployment state."
    )


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--minimum-gib", type=float, default=DEFAULT_MINIMUM_GIB)
    parser.add_argument("--path", type=Path, default=Path.cwd())
    args = parser.parse_args()
    if not math.isfinite(args.minimum_gib) or args.minimum_gib <= 0:
        parser.error("--minimum-gib must be greater than zero")

    free_bytes = shutil.disk_usage(args.path).free
    minimum_bytes = int(args.minimum_gib * GIB)
    error = evaluate_headroom(free_bytes, minimum_bytes)
    if error:
        print(f"error: {error}")
        return 1
    print(
        "capacity preflight passed: "
        f"{free_bytes / GIB:.2f} GiB available; "
        f"{args.minimum_gib:.2f} GiB required"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
